### DISCUSSION TITLE
BIRT importPackage works in Eclipse Studio but not when deployed

### DISCUSSION THREAD

**Discussion by Planet-RWehner** [eclipse-birt/birt#970] [General]
Eclipse: Version: Neon.3 Release (4.6.3) Build id: 20170314-1500
BIRT: 4.6.0.v201606072122

Using BIRT rptdesign file in eclipse studio, I am using importPackage on a .jar file I created from an external library cronUtils.  Placed in subfolder to my main folder named 'lib'.   Included in report's Resources as 'lib/crontils.jar'. 
Folders are like:

.\(main report .js and .rptlibrary files)
.\reports\(the .rptdesign file)
.\lib\cronutils.jar
reporting engine uses .\ as main 'current' folder.
Running report in eclipse studio works fine with built in Viewer or to PDF.

Deployed to my server, same folder structure and files, when first trying to use one of the classes imported from the package I get an error message:  ReferenceError: "CronParser" is not defined.

Code like this:
report 'initialize' script

try {
    IRMNBirtlog("report 'initialize' - importing packages...", Level.INFO);  // just a logging function to a text file.
    importPackage(Packages.java.util.logging);
    importPackage(Packages.java.util); // for Locale
    importPackage(Packages.com.cronutils.parser);
    importPackage(Packages.com.cronutils.descriptor);
    importPackage(Packages.com.cronutils.model);
    importPackage(Packages.com.cronutils.model.definition);
} catch(e) {
    IRMNBirtlog("report 'initialize' - error importing packages: " + e, Level.INFO);
}

report 'open' script
IRMNBirtlog("init CronExpressionParser.", Level.INFO);
var theparser = new CronParser(CronDefinitionBuilder.instanceDefinitionFor(CronType.QUARTZ));

since it is running on the server, there is no call stack crash log, or I just don't know where to look for one.

Is there a way to find out if the 'importPackage' actually succeeded?  and if not, why?
Why would it work in the eclipse studio in either debug or normal perspective and not on server with same files.
I verified all files are on the server in correct structure.

in desperation I tried copying the .jar file to the same folder as the .rptdesign file and to the .\ folder with no success.
Thanks for any guidance you can provide.

**Comment by hvbtup** [comment]
First of all, you should give BIRT 4.9 a try.
I also often use Java libraries from within Javascript in BIRT.
I don't know what you mean with "Eclipse Studio".
Here's how it works for me.

### For development, I use BIRT Report Designer All-In-One.

External _Javascript_ libraries have to be in the directory which is configured here:
![grafik](https://user-images.githubusercontent.com/7644776/168852487-ae4a9223-e77f-462a-8df3-59d9afd1cd04.png)

![grafik](https://user-images.githubusercontent.com/7644776/168852680-bce37ade-5746-4501-8114-af4ca747708a.png)

That's the same directory as for *.rptlibrary files.

External _Java_ libraries which you want to access from Javascript have to be specified here:
![grafik](https://user-images.githubusercontent.com/7644776/168854091-c8bd5ea7-b17f-4b37-85ec-68cd896f79ab.png)

### At runtime, I am using the POJO runtime (with a custom integration, not the WebViewer).

In this case, it is as simple as: ReportEngine/lib and the external Java libraries which you want to access from Javascript have to be in the ClassPath. E.g. on Windows this might look like this:

    SET CLASSPATH=E:\jlib\birt\ReportEngine\lib\*;E:\jlib\tp\BIRT_Server\*;E:\jlib\tp\jark\*;E:\jlib\ida\*;E:\jlib\zxing\*;E:\instantclient\ojdbc7.jar

The syntax with "*" is not so well-known. It means: Add every file in this directory to the ClassPath (the directory should contain only *.jar files).

**Reply by Planet-RWehner** [reply]
Thanks for the quick response.

What I'm confused about is why the Eclipse Preferences has anything to do with the report execution.
When I copy files to my server's runtime environment, I copy only the .rptDesign file, my .rptLibrary file, my .js source and this particular .jar file, that is all.

And for that matter, I do not have any of those items set (intentionally) in preferences, because they are specifically added to the .rptDesign file itself in its own properties.   Note all reports (.rptdesign files) are in subfolders to where reporting root folder is.
![image](https://user-images.githubusercontent.com/105727637/168892502-8591758d-2391-45cf-bb82-63843dc44826.png)

I always get this error when trying to access a class from the .jar file
TypeError: Cannot call property instanceDefinitionFor in object [JavaPackage com.cronutils.model.definition.CronDefinitionBuilder]. It is not a function, it is "object".

I wonder if ANYONE has integrated Cron Expression utilities into a BIRT report, to get a Human readable form of an expression, and how?

**Reply by hvbtup** [reply]
> What I'm confused about is why the Eclipse Preferences has anything to do with the report execution.

The Eclipse Preferences are only used when you run the report directly from the IDE.

As I said before, at runtime (assuming that you are using the POJO runtime) the *.jar files have to be part of the Java ClassPath.
You have to configure the ClassPath yourself. There is no magic.

**Reply by Planet-RWehner** [reply]
Thanks for the information, I'll investigate further, but cannot find how to add 'classpath' to an application running on a server.  It doesn't know about drive letters or the Operating System, so this may be difficult.   
Disconcerting how it runs in Eclipse IDE, specified in the report design, but when moved outside it does not.  No special classpath specified in IDE.

**Reply by hvbtup** [reply]
> ... cannot find how to add 'classpath' to an application running on a server.

This is required knowledge for the developers and/or administrators of the software running on the server.
